src/main/java/io/github/hotlava03/chatutils/listeners/RetrieveChatListener.java

```diff
@@ -2,39 +2,68 @@
     @Override
     public void accept(JoinServerEvent e) {
-        try {
-            var client = MinecraftClient.getInstance();
-            var serverInfo = client.getCurrentServerEntry();
-            var address = serverInfo != null ? serverInfo.address : ChatStorage.SINGLEPLAYER_ADDRESS;
-            var storage = ChatStorage.getInstance();
-            storage.setBlockingChatEvents(true);
-
-            var lines = new ArrayList<>(storage.getStoredLines(address));
-            if (lines.isEmpty()) {
-                storage.setBlockingChatEvents(false);
-                return;
-            }
-            var date = new Date(storage.getTimestamp(address));
-
-            lines.forEach((line) -> client.inGameHud.getChatHud().addMessage(Text.Serializer.fromJson(line)));
+        var client = MinecraftClient.getInstance();
+        var serverInfo = client.getCurrentServerEntry();
+        var address = serverInfo != null ? serverInfo.address : null;
+        if (address == null) {
+            LogManager.getLogger().error("Unexpected missing server info for joined server.");
             client.inGameHud.getChatHud().addMessage(Text.literal(
-                    StringUtils.translateAlternateColorCodes(Text.translatable("chat-utils.stored_messages", date).getString())
+                    StringUtils.translateAlternateColorCodes("&c[CHAT UTILS] &4Unexpected: Missing server/world info upon server/world join.")
             ));
+            return;
+        }
+
+        var storage = ChatStorage.getInstance();
+        storage.setBlockingChatEvents(true);
+
+        // Chat Persist.
+        if (ChatUtilsConfig.ENABLE_CHAT_PERSIST.value()) {
+            handleChatPersist(storage, address, client);
+        }
+
+        // Command Persist.
+        if (ChatUtilsConfig.ENABLE_COMMAND_PERSIST.value()) {
+            handleCommandPersist(storage, address, e.getMessageHistory());
+        }
+
+        storage.setBlockingChatEvents(false);
+    }
+
+    private void handleChatPersist(ChatStorage storage, String address, MinecraftClient client) {
+        var chatLines = new ArrayList<>(storage.getStoredChatLines(address));
+        if (chatLines.isEmpty()) {
+            storage.setBlockingChatEvents(false);
+            return;
+        }
+        var date = new Date(storage.getTimestamp(address));
+
+        chatLines.forEach((line) -> client.inGameHud.getChatHud().addMessage(Text.Serializer.fromJson(line)));
+        client.inGameHud.getChatHud().addMessage(Text.literal(
+                StringUtils.translateAlternateColorCodes(Text.translatable("chat-utils.stored_messages", date).getString())
+        ));
+    }
+
+    private void handleCommandPersist(ChatStorage storage, String address, List<String> messageHistory) {
+        var cmdLines = new ArrayList<>(storage.getStoredCmdLines(address));
+        if (cmdLines.isEmpty()) {
             storage.setBlockingChatEvents(false);
```

Comment:
do chat events need to be blocked here, you're not modifying the HUD
just block and unblock in `handleChatPersist`
right and you're unblocking a 2nd time in the `accept` function anyway